### PR TITLE
don't warn of invalid price fields on unused price sets

### PR DESCRIPTION
**Pull request by MegaphoneJon**
Overview
----------------------------------------
There's an "invalid price fields" check that detects price fields using inactive/deleted financial types.  However, when disabling/deleting a financial type, this can generate a very large list of invalid price fields, many of which require no attention because they're unused.

Before
----------------------------------------
Flag any price field as invalid if the financial type is deleted/disabled.

After
----------------------------------------
Only flag price fields if the financial type is deleted/disabled, *and* the price field is used on an active event or contribution page.

Technical Details
----------------------------------------
This isn't as big of a change as it looks, but I had to move initialization of `$count`, `$html` and `$messages` to the top of the function.

After that, I just wrapped the SQL statement in a `foreach` so it's run twice - in each case joining to the contribution page or event tables to check if the public page is active.  Also to ensure the price field is in use at all (hence the use of inner joins).

Comments
----------------------------------------
This reduces the UX issue this check generates.  However, it would be reduced further if
* it were possible to mark a contribution page inactive from the "Manage Contribution Pages" page (from the "more..." menu, like Events).
* We also checked that the default financial types for contribution pages, events and membership types were all active.

These are out of scope for this change but aren't bad ideas.

### CODE REVIEW

**Comment by MegaphoneJon** [comment]
Another potential follow-on improvement is to add a "Used On" column to this check, which links to the event/contribution page it's used on.  It's frequently the case that these can be disabled.

**Comment by seamuslee001** [comment]
Jenkins retest this please

**Comment by MegaphoneJon** [comment]
Pretty sure this fail is unrelated.

File: CRM/Utils/Check/Component/PriceFields.php [hunk 1]
```diff
@@ -22,24 +22,28 @@ class CRM_Utils_Check_Component_PriceFields extends CRM_Utils_Check_Component {
...
-      ]);
-      $html .= "<tr><td>$dao->ps_title</td><td>$dao->psf_label</td><td><a href='$url'>" . ts('View Price Set Fields') . '</a></td></tr>';
+    foreach (['civicrm_contribution_page', 'civicrm_event'] as $publicPage) {
+      $sql = "SELECT DISTINCT ps.title as ps_title, ps.id as ps_id, psf.label as psf_label
+        FROM civicrm_price_set ps
+        INNER JOIN civicrm_price_field psf ON psf.price_set_id = ps.id
+        INNER JOIN civicrm_price_field_value pfv ON pfv.price_field_id = psf.id
+        LEFT JOIN civicrm_financial_type cft ON cft.id = pfv.financial_type_id
+        JOIN civicrm_price_set_entity pse ON pse.entity_table = '$publicPage' AND pse.price_set_id = ps.id
+        JOIN $publicPage pp ON pse.entity_id = pp.id ANd pp.is_active = 1
```
Comment (larssandergreen) [inline CRM/Utils/Check/Component/PriceFields.php:35]:
```suggestion
        JOIN $publicPage pp ON pse.entity_id = pp.id AND pp.is_active = 1
```

**Comment by larssandergreen** [comment]
This is definitely an annoying and unhelpful warning. However, a contribution price set can still be used for a contribution from the back end, even if it is not used on an active contribution page. So it does seem reasonable to warn for any contribution price set, even if it is kind of annoying. I think you'd also need to cover membership price sets here too, which can also be used on the back end even if not in use on a contribution page.

For events, what about Change Selections? I can still do Change Selections on a participant and change their price set selections for an event that isn't active. If I change their selection to a price field that has an inactive or deleted financial type, what happens? Not sure, but it seems like something might break.

I also thought it would make sense to check if the price set itself is active and the price field itself is active, but this would also be an issue when editing a contribution or participant (as you can still change selections even when the price set isn't active).

So I'm not sure what the best way forward here is. Better to have more warnings? Or better to not show warnings for price sets only used in inactive events, because users are unlikely to change the selections for inactive events? Similar for inactive price sets or fields.

Test fail is definitely unrelated.

**Comment by eileenmcnaughton** [comment]
@MegaphoneJon this feels familiar - it wasn't otherwise merged was it?

**Comment by MegaphoneJon** [comment]
@eileenmcnaughton Yes - I wish I could remember these things better.  But we just merged https://github.com/civicrm/civicrm-core/pull/28322, which was a fix on the same issue.

Comparing the PRs - this one is mildly cleverer but the other is better for performance.  Let's close this.

**Pull request closed without merging**